### ISSUE TITLE
Inadvertently removed the `buffer/asBuffer` optimization

### ISSUE THREAD

**Issue by vweevers** [Level/levelup#515]
Continuing from https://github.com/Level/encoding-down/pull/17#issuecomment-337314850. I found this issue trying to make sense of the `buffer` option on encodings, and eventually traced it back to `levelup`.

In `levelup@1`, we had an optimization for string-based encodings like [`json`](https://github.com/Level/codec/blob/cb7eac67803c5a3af0ef998721c6b9317fdfada4/lib/encodings.js#L16):

```js
exports.json = {
  encode: JSON.stringify,
  decode: JSON.parse,
  buffer: false,
  type: 'json'
}
```

The important property here is `buffer: false`. This told levelup to prefer to fetch a string from the underlying store, rather than a Buffer. The underlying store would receive `{ asBuffer: false }` in `get` options and similar places.

This is how the code looked:

```js
LevelUP.prototype.get = function (key_, options, callback) {
  options.asBuffer = this._codec.valueAsBuffer(options)

  this.db.get(key, options, function (err, value) {}
}
```

When we removed the encoding logic from `levelup` (#458), the `buffer/asBuffer` logic was [also removed](https://github.com/Level/levelup/pull/458/files#diff-f3f1287944486f89a62e35c994ebca54L173) and did not get a replacement elsewhere.

So since `levelup@2`, underlying stores always see `{ asBuffer: true }`. The means we incur a conversion cost, sometimes even double conversion:

- `leveldown` always returns keys and values as Buffers (with [an additional memcpy cost](https://github.com/Level/leveldown/commit/b7272db9f195a1070f4cdd15a99216ddc08f86eb#diff-0a0a90a8810da7685a5be3a1657da592R134), but let's leave that aside for a moment)
- `memdown` as well will convert strings to Buffers (if they were stored as strings, which is true for the `json` encoding at least)
-  The encodings convert Buffers back to strings (in the case of `json` it implicitly does `JSON.parse(x.toString())`).

**Comment by vweevers** [comment]
@ralphtheninja @juliangruber

**Comment by juliangruber** [comment]
Since this is about encodings, and needs to know about encodings, can we solve this in encoding-down? This part of the stack knows which option is best, and should then be able to pass it to the underlying store.

Great find, @vweevers!

**Comment by ralphtheninja** [comment]
> Since this is about encodings, and needs to know about encodings, can we solve this in encoding-down? This part of the stack knows which option is best, and should then be able to pass it to the underlying store.

Sounds like a good place to me!

@vweevers Really awesome that you found this! Seems to be me that messed it up. Before we decide and implement a fix, it would be nice to have some failing tests for this behavior.

**Comment by vweevers** [comment]
> Before we decide and implement a fix, it would be nice to have some failing tests for this behavior.

👍 I'll rewrite what I have into a test for `encoding-down`.

**Comment by vweevers** [comment]
This was fixed in Level/encoding-down#19, released as patch (`encoding-down@2.3.4`), which automatically rolls into `level-packager` and hence `level`.
